providers/implementations/keymgmt/mlx_kmgmt.c

```diff
@@ -0,0 +1,764 @@
+                              const OSSL_PARAM params[])
+{
+    static const int minimal =
+        OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS
+        | OSSL_KEYMGMT_SELECT_PRIVATE_KEY;
+    PROV_ML_KEM_GEN_CTX *gctx = NULL;
+
+    /*
+     * We can only generate private keys, check that the selection is
+     * appropriate.
+     */
+    if (!ossl_prov_is_running()
+        || (selection & minimal) == 0
+        || (gctx = OPENSSL_zalloc(sizeof(*gctx))) == NULL)
+        return NULL;
+
+    gctx->variant = v;
+    gctx->libctx = libctx;
+    gctx->selection = selection;
+    if (mlx_kem_gen_set_params(gctx, params))
+        return gctx;
+
+    mlx_kem_gen_cleanup(gctx);
+    return NULL;
+}
+
+static const OSSL_PARAM *mlx_kem_gen_settable_params(ossl_unused void *vgctx,
+                                                     ossl_unused void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PROPERTIES, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return settable;
+}
+
+static void *mlx_kem_gen(void *vgctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+    MLX_KEY *key;
+    char *propq = gctx->propq;
+
+    if (gctx == NULL
+        || (gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) ==
+            OSSL_KEYMGMT_SELECT_PUBLIC_KEY)
+        return NULL;
+
+    /* Lose ownership of propq */
+    gctx->propq = NULL;
+    if ((key = mlx_kem_key_new(gctx->variant, gctx->libctx, propq)) == NULL)
+        return NULL;
+
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return key;
+
+    /* XXX: Using same "propq" for all components */
+    key->mkey = EVP_PKEY_Q_keygen(gctx->libctx, gctx->propq,
+                                  key->minfo->algorithm_name);
+    key->xkey = EVP_PKEY_Q_keygen(gctx->libctx, gctx->propq,
+                                  key->xinfo->algorithm_name,
+                                  key->xinfo->group_name);
+    if (key->mkey != NULL && key->xkey != NULL) {
+        key->state = MLX_HAVE_PRVKEY;
+        return key;
+    }
+
+    mlx_kem_key_free(key);
+    return NULL;
+}
+
+static void mlx_kem_gen_cleanup(void *vgctx)
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+
+    if (gctx == NULL)
+        return;
+    OPENSSL_free(gctx->propq);
+    OPENSSL_free(gctx);
+}
+
+static void *mlx_kem_dup(const void *vkey, int selection)
+{
+    const MLX_KEY *key = vkey;
+    MLX_KEY *new;
+
+    if (!ossl_prov_is_running()
+        || (new = OPENSSL_memdup(key, sizeof(*new))) == NULL)
+        return NULL;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0) {
+        new->xkey = new->mkey = NULL;
+        return new;
+    }
+
+    new->mkey = EVP_PKEY_dup(key->mkey);
+    new->xkey = EVP_PKEY_dup(key->xkey);
+    if (new->xkey != NULL && new->mkey != NULL)
+        return new;
+
+    mlx_kem_key_free(new);
+    return NULL;
+}
+
+typedef void (*func_ptr_t)(void);
```

Comment:
This screams out to be a defined OpenSSL typedef ... which could then be more widely used.
Could even be called OSSL_FUNC_PTR ... or even OSSL_FUNC if you want to save a character ...